ami/trainers/hifigan_trainer.py

```diff
@@ -0,0 +1,393 @@
+                loss_adv_g = self._generator_adversarial_losses(
+                    authenticity_list_mpd_fake
+                ) + self._generator_adversarial_losses(authenticity_list_msd_fake)
+                # calc sum
+                loss_g = loss_rec + loss_fm + loss_adv_g
+                optimizer_g.zero_grad()
+                loss_g.backward()
+                optimizer_g.step()
+                self.logger.log(self.log_prefix + "losses/generator/train-reconstruction", loss_rec)
+                self.logger.log(self.log_prefix + "losses/generator/train-feature_matching", loss_fm)
+                self.logger.log(self.log_prefix + "losses/generator/train-adversarial", loss_adv_g)
+
+                self.logger.update()
+
+        if self.validation_dataloader is not None:
+            self.validation(self.validation_dataloader)
+
+        self.optimizer_state_g = optimizer_g.state_dict()
+        self.optimizer_state_mpd = optimizer_mpd.state_dict()
+        self.optimizer_state_msd = optimizer_msd.state_dict()
+        self.logger_state = self.logger.state_dict()
+
+    @override
+    def save_state(self, path: Path) -> None:
+        path.mkdir()
+        torch.save(self.optimizer_state_g, path / "optimizer_g.pt")
+        torch.save(self.optimizer_state_mpd, path / "optimizer_mpd.pt")
+        torch.save(self.optimizer_state_msd, path / "optimizer_msd.pt")
+        torch.save(self.logger.state_dict(), path / "logger.pt")
+        torch.save(self.dataset_previous_get_time, path / "dataset_previous_get_time.pt")
+
+    @override
+    def load_state(self, path: Path) -> None:
+        self.optimizer_state_g = torch.load(path / "optimizer_g.pt")
+        self.optimizer_state_mpd = torch.load(path / "optimizer_mpd.pt")
+        self.optimizer_state_msd = torch.load(path / "optimizer_msd.pt")
+        self.logger.load_state_dict(torch.load(path / "logger.pt"))
+        self.dataset_previous_get_time = torch.load(path / "dataset_previous_get_time.pt")
+
+    def _discriminator_adversarial_losses(
+        self, authenticity_list_real: list[torch.Tensor], authenticity_list_fake: list[torch.Tensor]
+    ) -> torch.Tensor:
+        loss = torch.tensor(0.0).type_as(authenticity_list_real[0])
+        for authenticity_real, authenticity_fake in zip(authenticity_list_real, authenticity_list_fake):
+            real_loss = torch.mean((1 - authenticity_real) ** 2)
+            fake_loss = torch.mean(authenticity_fake**2)
+            loss += real_loss + fake_loss
+        return loss
+
+    def _feature_loss(
+        self, fmaps_list_real: list[list[torch.Tensor]], fmaps_list_fake: list[list[torch.Tensor]]
+    ) -> torch.Tensor:
+        loss = torch.tensor(0.0).type_as(fmaps_list_real[0][0])
+        for fmaps_real, fmaps_fake in zip(fmaps_list_real, fmaps_list_fake):
+            for fmap_real, fmap_fake in zip(fmaps_real, fmaps_fake):
+                loss += torch.mean(torch.abs(fmap_real - fmap_fake))
```

Comment:
名前付のハミング距離が近いためtypoの危険がある。
```suggestion
        for fmaps_real, fmaps_fake in zip(fmaps_list_real, fmaps_list_fake):
            for fm_real, fm_fake in zip(fmaps_real, fmaps_fake):
                loss += torch.mean(torch.abs(fm_real - fm_fake))
```